### PR TITLE
Add Camera Cycle to camera plugin

### PR DESCRIPTION
**Pull request by hjdarnel**
The idea with this functionality is that multiple repeated clicks would cycle the camera through the cardinal directions. Two quick clicks would reset to Look South, three to Look East, etc. The timeout that determines how quickly these clicks need to be is configurable (mouse vs tablet users could enjoy different values)

This is off by default. It shouldn't conflict with the Compass options `compassLook` functionality, but just extend it. The order will continue to be `North, South, East, West` however if East is the current top menu entry the menu will show `East, West, North, South` to prevent duplicate menu entries.

Closes https://github.com/runelite/runelite/issues/10731 particularly this comment https://github.com/runelite/runelite/issues/10731#issuecomment-582265261

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java [hunk 1]
```diff
@@ -117,36 +127,83 @@ protected void shutDown()
...
+				menuEntries = Arrays.copyOf(menuEntries, menuEntries.length + 3);
+
+				// fill the menu from top to bottom
+				// but iterate through options cyclically from 0 -> 3 -> 0
+				int entryIndex = menuEntries.length - 1;
+
+				// The handling for these entries is done in ToplevelCompassOp.rs2asm`
+				menuEntries[entryIndex] = createCameraLookEntry(menuEntryAdded, currentOptionIndex + 1, options[currentOptionIndex % count]);
+				entryIndex -= 1;
+				currentOptionIndex += 1;
```
Comment (hjdarnel) [inline runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java]: Lines 164-165 could be removed with something like `menuEntries[entryIndex--]` and `currentOptionIndex++` but I found it hard to follow when written that way

File: runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraConfig.java [hunk 2]
```diff
@@ -158,4 +159,27 @@ default boolean compassLook()
...
+	}
+
+	@ConfigItem(
+		keyName = "compassMultipleClickTimeout",
+		name = "Multiple click timeout",
+		description = "Configure how quickly the compass resets to \"Look North\" after the last click",
+		position = 12
+	)
+	@Units(Units.SECONDS)
+	default int compassMultipleClickTimeout()
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraConfig.java]: This currently isn't used. There's a `RESET_DIRECTION_TIMEOUT` constant which should have its usages replaced by this.

Comment (hjdarnel) [reply runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraConfig.java]: My bad, you're right. Fixed.

File: runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java [hunk 3]
```diff
@@ -66,6 +71,11 @@
 	private static final String LOOK_SOUTH = "Look South";
 	private static final String LOOK_EAST = "Look East";
 	private static final String LOOK_WEST = "Look West";
+	private static final int RESET_DIRECTION_TIMEOUT = 5;
+	private int lastUpdatedTick = 0;
+	private static final String defaultLeftClickOption = LOOK_NORTH;
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java]: This should be moved with the other constants and renamed to use the same naming scheme.
```suggestion
	private static final String DEFAULT_LEFT_CLICK_OPTION = LOOK_NORTH;
```

File: runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java [hunk 4]
```diff
@@ -66,6 +71,11 @@
...
+	private String currentLeftClickOption;
+	private String[] options = new String[]{LOOK_NORTH, LOOK_SOUTH, LOOK_EAST, LOOK_WEST};
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java]: This can be made `static final` and moved with the other constants.
```suggestion
	private static final String[] OPTIONS = new String[]{LOOK_NORTH, LOOK_SOUTH, LOOK_EAST, LOOK_WEST};
```

Comment (Nightfirecat) [reply runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java]: Also NSEW is a strange order, I would expect to see this in NESW order.

File: runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java [hunk 5]
```diff
@@ -117,36 +128,84 @@ protected void shutDown()
...
+			int currentOptionIndex = ArrayUtils.indexOf(options, currentLeftClickOption);
 
-			// The handling for these entries is done in ToplevelCompassOp.rs2asm
-			menuEntries[--len] = createCameraLookEntry(menuEntryAdded, 4, LOOK_WEST);
-			menuEntries[++len] = createCameraLookEntry(menuEntryAdded, 3, LOOK_EAST);
-			menuEntries[++len] = createCameraLookEntry(menuEntryAdded, 2, LOOK_SOUTH);
-			menuEntries[++len] = north;
+			if (config.compassLook())
+			{
+				menuEntries = Arrays.copyOf(menuEntries, menuEntries.length + 3);
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java]: `3` should instead be a variable or reference to a variable, such as `OPTIONS.length - 1` since it makes no sense to be a magic number.

File: runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java [hunk 6]
```diff
@@ -117,36 +128,84 @@ protected void shutDown()
...
+				currentOptionIndex += 1;
+				menuEntries[entryIndex] = createCameraLookEntry(menuEntryAdded, currentOptionIndex + 1, options[currentOptionIndex % count]);
+				entryIndex -= 1;
+				currentOptionIndex += 1;
+				menuEntries[entryIndex] = createCameraLookEntry(menuEntryAdded, currentOptionIndex + 1, options[currentOptionIndex % count]);
+			}
+			else
+			{
+				// set top entry to the current left click
+				menuEntries[menuEntries.length - 1] = createCameraLookEntry(menuEntryAdded, currentOptionIndex + 1, options[currentOptionIndex % count]);
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java]: I don't understand... Why is `onMenuEntryAdded` running any code when `config.compassLook()` is false? We shouldn't be adding or changing any menu entries when that's the case.

File: runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java [hunk 7]
```diff
@@ -117,36 +128,84 @@ protected void shutDown()
...
+				menuEntries[entryIndex] = createCameraLookEntry(menuEntryAdded, currentOptionIndex + 1, options[currentOptionIndex % count]);
+				entryIndex -= 1;
+				currentOptionIndex += 1;
+				menuEntries[entryIndex] = createCameraLookEntry(menuEntryAdded, currentOptionIndex + 1, options[currentOptionIndex % count]);
+				entryIndex -= 1;
+				currentOptionIndex += 1;
+				menuEntries[entryIndex] = createCameraLookEntry(menuEntryAdded, currentOptionIndex + 1, options[currentOptionIndex % count]);
+				entryIndex -= 1;
+				currentOptionIndex += 1;
+				menuEntries[entryIndex] = createCameraLookEntry(menuEntryAdded, currentOptionIndex + 1, options[currentOptionIndex % count]);
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java]: This code is completely duplicated and should be replaced by a loop, such as:

```java
for (int i = 0; i < count; i++)
{
	menuEntries[menuEntries.length - i] = createCameraLookEntry(menuEntryAdded, currentOptionIndex + i, OPTIONS[currentOptionIndex % count]);
}
```

File: runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java [hunk 8]
```diff
@@ -117,36 +128,84 @@ protected void shutDown()
...
 			client.setMenuEntries(menuEntries);
 		}
 	}
 
-	private MenuEntry createCameraLookEntry(MenuEntryAdded lookNorth, int identifier, String option)
+	@Subscribe
+	private void onMenuOptionClicked(MenuOptionClicked event)
+	{
+		// if the option is disabled, don't process the click any differently
+		if ((event.getMenuAction() == MenuAction.CC_OP) && config.compassMultipleClickCycle() && event.getMenuOption().equals(currentLeftClickOption))
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java]:
```suggestion
		if (event.getMenuAction() == MenuAction.CC_OP && config.compassMultipleClickCycle() && event.getMenuOption().equals(currentLeftClickOption))
```

File: runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java [hunk 9]
```diff
@@ -117,36 +128,84 @@ protected void shutDown()
...
 
 			client.setMenuEntries(menuEntries);
 		}
 	}
 
-	private MenuEntry createCameraLookEntry(MenuEntryAdded lookNorth, int identifier, String option)
+	@Subscribe
+	private void onMenuOptionClicked(MenuOptionClicked event)
+	{
+		// if the option is disabled, don't process the click any differently
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java]: Do we have a way to tell whether this menu option was clicked via left-clicking the map or clicking a right-click menu? (regardless of if it was the top menu option) I would think that the cycling behavior would not be wanted if clicking a right-click menu option.

File: runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java (hunk 4 above)
Comment (hjdarnel) [reply runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java]: That's the order that the current Camera Options plugin shows options, so I didn't want to do different than that. I can change this to NESW though.

File: runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java (hunk 6 above)
Comment (hjdarnel) [reply runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java]: if `config.compassLook() == true` then we need to build out the menu options for that plugin, and rearrange them depending on the current left click option. If `config.compassLook() == false` then we still need to modify the top option to Look South, East, West, etc depending on the current left click value. 

If `config.compassMultipleClickTimeout() == false` (which is the config option I'm adding) then we just don't do anything special on a compass click, instead `currentLeftClickOption` will always be the default "Look North"

File: runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java (hunk 4 above)
Comment (Nightfirecat) [reply runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java]: Hm, that's fine then. Let's leave it be.

**Comment by hjdarnel** [comment]
Should this just be an external plugin?

**Comment by Nightfirecat** [comment]
Based on discussion in Discord, since the "Look East/South/West" feature already exists in the main client, this seems reasonable to go into the main client as well.

File: runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java (hunk 9 above)
Comment (hjdarnel) [reply runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java:232]: this is covered by line 233 `event.getMenuOption().equals(currentLeftClickOption)` I think.

**Comment by hjdarnel** [comment]
After a long hiatus... this PR can be reviewed again

**Comment by Nightfirecat** [comment]
Needs rebase now that Look NESW is now a vanilla feature and it was removed from the plugin.

**Comment by hjdarnel** [comment]
I'm having trouble rebasing this. I don't really understand how the menu entry class works. If someone else wanted to pick this up and put their own PR up they can feel free.
